File: src/test/kotlin/com/vadimkomis/localization/LocaleManifestTests.kt
package com.vadimkomis.localization

import com.google.gson.Gson
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

private data class LocaleManifest(
    val sourceLocale: String,
    val targetLocales: List<String>
)

class LocaleManifestTests {
    private fun loadManifest(): LocaleManifest {
        val path = System.getProperty("locales.manifest") ?: "locales.json"

        return Gson().fromJson(File(path).readText(), LocaleManifest::class.java)
    }

    @Test
    fun `source locale matches manifest`() {
        val manifest = loadManifest()

        assertEquals(manifest.sourceLocale, SupportedLocale.sourceLocale.identifier)
    }

    @Test
    fun `target locales match manifest order`() {
        val manifest = loadManifest()

        assertEquals(manifest.targetLocales, SupportedLocale.targetLocales.map { it.identifier })
    }

    @Test
    fun `every manifest identifier resolves to a supported locale`() {
        val manifest = loadManifest()
        val identifiers = listOf(manifest.sourceLocale) + manifest.targetLocales

        for (identifier in identifiers) {
            assertNotNull(SupportedLocale.fromIdentifier(identifier), "Unknown identifier: $identifier")
        }
    }
}

File: Tests/LocalizationTests/LocaleManifestTests.swift
import Foundation
import Testing
@testable import Localization

private struct LocaleManifest: Decodable {
    let sourceLocale: String
    let targetLocales: [String]
}

@Suite("Locale manifest sync")
struct LocaleManifestTests {
    private func loadManifest() throws -> LocaleManifest {
        let manifestURL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("locales.json")
        let data = try Data(contentsOf: manifestURL)

        return try JSONDecoder().decode(LocaleManifest.self, from: data)
    }

    @Test("source locale matches locales.json manifest")
    func sourceLocaleMatchesManifest() throws {
        let manifest = try loadManifest()

        #expect(manifest.sourceLocale == SupportedLocale.sourceLocale.identifier)
    }

    @Test("target locales match locales.json manifest order")
    func targetLocalesMatchManifest() throws {
        let manifest = try loadManifest()

        #expect(manifest.targetLocales == SupportedLocale.targetLocales.map(\.identifier))
    }

    @Test("every manifest identifier resolves to a supported locale")
    func manifestIdentifiersResolve() throws {
        let manifest = try loadManifest()
        let identifiers = [manifest.sourceLocale] + manifest.targetLocales

        for identifier in identifiers {
            #expect(SupportedLocale(rawValue: identifier) != nil)
        }
    }
}
